src/app/data-model-provider/MetadataList.h

```diff
@@ -0,0 +1,176 @@
+class MetadataList
+{
+public:
+    using SpanType = Span<const T>;
+
+    MetadataList()                                       = default;
+    MetadataList(const MetadataList &)                   = delete;
+    MetadataList & operator=(const MetadataList & other) = delete;
+
+    MetadataList(MetadataList && other) { *this = std::move(other); }
+
+    MetadataList & operator=(MetadataList && other)
+    {
+        if (this != &other)
+        {
+            this->mBackingStorage.Free();
+
+            this->mBackingStorage = std::move(other.mBackingStorage);
+            this->mSize           = other.mSize;
+            this->mIsImmutable    = other.mIsImmutable;
+            this->mSpan           = other.mSpan;
+
+            other.mIsImmutable = true;
+            other.mSize        = 0;
+            other.mSpan        = Span<T>{};
+        }
+        return *this;
+    }
+
+    const T & operator[](size_t index)
+    {
+        VerifyOrDie((mSize > 0) && (index < mSize));
+
+        if (this->mBackingStorage)
+        {
+            return Span<const T>{ mBackingStorage.Get(), mSize }[index];
+        }
+
+        return mSpan[index];
+    }
+
+    template <size_t N>
+    static MetadataList FromArray(const std::array<T, N> & arr)
+    {
+        MetadataList<T> list;
+
+        if (list.reserve(N) != CHIP_NO_ERROR)
+        {
+            list.Invalidate();
+            return list;
+        }
+
+        for (const auto & item : arr)
+        {
+            if (list.Append(item) != CHIP_NO_ERROR)
+            {
+                list.Invalidate();
+                return list;
+            }
+        }
+
+        list.mIsImmutable = true;
+        return list;
+    }
+
+    template <size_t N>
+    static MetadataList FromConstArray(const std::array<const T, N> & arr)
+    {
+        MetadataList<T> list;
+        list.mSpan        = Span<const T>{ arr.data(), arr.size() };
+        list.mSize        = arr.size();
+        list.mIsImmutable = true;
+
+        return list;
+    }
+
+    static MetadataList FromConstSpan(const Span<const T> & span)
+    {
+        MetadataList<T> list;
+        list.mSpan        = span;
+        list.mSize        = span.size();
+        list.mIsImmutable = true;
+
+        return list;
+    }
+
+    CHIP_ERROR reserve(size_t numElements)
+    {
+        VerifyOrReturnError(!mIsImmutable, CHIP_ERROR_INCORRECT_STATE);
+        VerifyOrReturnError(mSize == 0, CHIP_ERROR_INCORRECT_STATE);
+        if (!mBackingStorage.Calloc(numElements))
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        return CHIP_NO_ERROR;
+    }
+
+    size_t size() const { return mSize; }
+
+    bool empty() const { return size() == 0u; }
+
+    size_t capacity() const { return std::max(mSize, mBackingStorage.AllocatedSize()); }
+
+    void Invalidate()
+    {
+        mBackingStorage.Free();
+        mSize        = 0;
+        mIsImmutable = true;
+    }
+
+    Span<const T> GetSpanValidForLifetime() const
```

Comment:
If we're returning it by value, why do we need to set mSpan at all?  And if we stopped doing that, we would not need `mSpan` to be mutable....

Reply:
I removed the mSpan. I believe this may have been an intermediate to allow both dynamic and const usage added after the fact.